### ISSUE TITLE
Wrong type inference for closure without explicit return type

### ISSUE THREAD

**Issue by liugangnhm** [intellij-rust/intellij-rust#8892]
## Environment

* **IntelliJ Rust plugin version:** 0.4.171.4656-221
* **Rust toolchain version:** 1.61.0 (fe5b13d68 2022-05-18) x86_64-pc-windows-msvc
* **IDE name and version:** GoLand 2022.1 (GO-221.5080.224)
* **Operating system:** Windows 10 10.0
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description
![image](https://user-images.githubusercontent.com/743845/171305880-307caa1c-0546-4a8b-ae90-a6d5e19bde0e.png)

```rust
fn main() {
    fn function(i: i32) -> i32 { i + 1 }

    let closure_annotated = |i: i32| -> i32 { i + 1 };
    let closure_inferred = |i| i + 1;

    let i = 1;
    // 调用函数和闭包。
    println!("function: {}", function(i));
    println!("closure_annotated: {}", closure_annotated(i));
    println!("closure_inferred: {}", closure_inferred(i));

    // 没有参数的闭包，返回一个 `i32` 类型。
    // 返回类型是自动推导的。
    let one = || 1;
    println!("closure returning one: {}", one());
}
```

## Steps to reproduce

copy above code to your rust project

**Comment by Undin** [comment]
@vlad20012 it looks similar to one case described in #8835, but it still doesn't work as expected

**Comment by YthanZhang** [comment]
This code doesn't produce the false positive

```rust
fn simple_closure() {
    let return_it = |x| -> _ { x };

    let s = return_it("test string");

    println!("{}", s);
}
```

But if I remove the `-> _` it will

**Comment by AdamLeyshon** [comment]
```
CLion 2022.2.3
Build #CL-222.4167.35, built on September 15, 2022
Licensed to Adam Leyshon
Subscription is active until October 18, 2022.
Runtime version: 17.0.4+7-b469.53 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Linux 5.15.60-1-manjaro
GC: G1 Young Generation, G1 Old Generation
Memory: 8096M
Cores: 24
Registry:
    debugger.watches.in.variables=false
    run.processes.with.pty=TRUE

Non-Bundled Plugins:
    org.rust.lang (0.4.179.4903-222)
    net.sf.opk.avro-schema-support (213.2.1)
    idea.plugin.protoeditor (222.4167.21)
    com.hiberbee.intellij.hiberbee-theme (2022.5.16.1805)
    Key Promoter X (2022.3)
    mobi.hsz.idea.gitignore (4.4.2)
    com.almightyalpaca.intellij.plugins.discord (1.10.0)
    ru.adelf.idea.dotenv (2022.2)

Current Desktop: KDE
```

This is still an issue for many types, `i32`, `i64`, unsigned, `String`, doesn't matter.
It gets worse if the type is behind some sort of Mutex or Cell.

For example I have a type wrapped in a OnceCell,

```
lazy_static! {
    pub static ref SETTINGS: OnceCell<Settings> = OnceCell::new();
}
```

And it can't autocomplete/infer the type or any of its methods/fields unless I invalidate the cache and restart.

Also for some reason in the latest builds it also seems to think I'm trying to move bool/numberic values whenever I try to compare them though I don't know if this is related to the inference engine problem.

![image](https://user-images.githubusercontent.com/7215574/194163859-9db970b0-b358-4b60-a7bc-3daeea738352.png)

![image](https://user-images.githubusercontent.com/7215574/194164637-a5050fc2-0746-4b44-9f30-7cf8d3771b70.png)

![image](https://user-images.githubusercontent.com/7215574/194164759-c3c2a606-1b01-478d-8626-faad0d00b761.png)

**Comment by Undin** [comment]
@liugangnhm your issue is not related to the original one

But I think it's the same as #9432 and it's already fixed by #9461. The fix will be landed in next release next week. Right now you may manually downgrade to `0.4.178` version by loading it from [Marketplace](https://plugins.jetbrains.com/plugin/8182-rust/versions/stable/222064) and [installing from disk](https://www.jetbrains.com/help/idea/managing-plugins.html#install_plugin_from_disk) or wait tomorrow nightly build and install [nightly](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-quick-start.html#install-nightly) plugin version
